src/scheduler/scheduler.rs

```diff
@@ -0,0 +1,318 @@
+use super::stat::SchedulerStat;
+use super::work::Work;
+use super::work_bucket::*;
+use super::worker::{Worker, WorkerGroup};
+use super::*;
+use crate::mmtk::MMTK;
+use crate::plan::Plan;
+use crate::util::OpaquePointer;
+use crate::vm::VMBinding;
+use std::collections::HashMap;
+use std::sync::atomic::Ordering;
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::{Arc, Condvar, Mutex};
+
+pub enum CoordinatorMessage<C: Context> {
+    Work(Box<dyn CoordinatorWork<C>>),
+    AllWorkerParked,
+    BucketDrained,
+}
+
+pub struct Scheduler<C: Context> {
+    /// Works that are scheduable at any time
+    pub unconstrained_works: WorkBucket<C>,
+    /// Works that are scheduable within Stop-the-world
+    pub prepare_stage: WorkBucket<C>,
+    pub closure_stage: WorkBucket<C>,
+    pub release_stage: WorkBucket<C>,
+    pub final_stage: WorkBucket<C>,
+    /// Works for the coordinator thread
+    pub coordinator_works: WorkBucket<C>,
+    /// workers
+    worker_group: Option<Arc<WorkerGroup<C>>>,
+    /// Condition Variable for worker synchronization
+    pub worker_monitor: Arc<(Mutex<()>, Condvar)>,
+    context: Option<&'static C>,
+    coordinator_worker: Option<Worker<C>>,
+    /// A message channel to send new coordinator works and other actions to the coordinator thread
+    pub channel: (
+        Sender<CoordinatorMessage<C>>,
+        Receiver<CoordinatorMessage<C>>,
+    ),
+    startup: Mutex<Option<Box<dyn CoordinatorWork<C>>>>,
+    finalizer: Mutex<Option<Box<dyn CoordinatorWork<C>>>>,
+}
+
+unsafe impl<C: Context> Send for Scheduler<C> {}
+unsafe impl<C: Context> Sync for Scheduler<C> {}
+
+impl<C: Context> Scheduler<C> {
+    pub fn new() -> Arc<Self> {
+        let worker_monitor: Arc<(Mutex<()>, Condvar)> = Default::default();
+        Arc::new(Self {
+            unconstrained_works: WorkBucket::new(true, worker_monitor.clone()), // `default_bucket` is always activated
+            prepare_stage: WorkBucket::new(false, worker_monitor.clone()),
+            closure_stage: WorkBucket::new(false, worker_monitor.clone()),
+            release_stage: WorkBucket::new(false, worker_monitor.clone()),
+            final_stage: WorkBucket::new(false, worker_monitor.clone()),
+            coordinator_works: WorkBucket::new(true, worker_monitor.clone()),
+            worker_group: None,
+            worker_monitor,
+            context: None,
+            coordinator_worker: None,
+            channel: channel(),
+            startup: Mutex::new(None),
+            finalizer: Mutex::new(None),
+        })
+    }
+
+    #[inline]
+    pub fn num_workers(&self) -> usize {
+        self.worker_group.as_ref().unwrap().worker_count()
+    }
+
+    pub fn initialize(
```

Comment:
post-alloc initialization is required for now because when MMTK<VM>/Scheduler instance is created, it may not get enough information it required to complete initialization. e.g. the number of worker threads. Not sure how can we fix this...